h2/src/main/org/h2/mvstore/Page.java

```diff
@@ -887,37 +898,50 @@ public int hashCode() {
         return pos != 0 ? (int) (pos | (pos >>> 32)) : super.hashCode();
     }
 
+    private boolean isPersistent() {
+        return memory != IN_MEMORY;
+    }
+
     public int getMemory() {
-        if (MVStore.ASSERT) {
-            int mem = memory;
-            recalculateMemory();
-            if (mem != memory) {
-                throw DataUtils.newIllegalStateException(
-                        DataUtils.ERROR_INTERNAL, "Memory calculation error");
+        if (isPersistent()) {
+            if (MVStore.ASSERT) {
+                int mem = memory;
+                recalculateMemory();
+                if (mem != memory) {
+                    throw DataUtils.newIllegalStateException(
+                            DataUtils.ERROR_INTERNAL, "Memory calculation error");
+                }
             }
+            return memory;
         }
-        return memory;
+        return getKeyCount();
     }
 
     private void addMemory(int mem) {
+        if(!isPersistent()) {
+            throw DataUtils.newIllegalStateException(
+                    DataUtils.ERROR_INTERNAL, "Memory calculation error2");
+        }
         memory += mem;
     }
 
     private void recalculateMemory() {
-        int mem = DataUtils.PAGE_MEMORY;
-        DataType keyType = map.getKeyType();
-        for (int i = 0; i < keys.length; i++) {
-            mem += keyType.getMemory(keys[i]);
-        }
-        if (this.isLeaf()) {
-            DataType valueType = map.getValueType();
+        if(isPersistent()) {
```

Comment:
should we not be throwing an exception here is we get called in non-persistent mode?